recipes/Media/ASR/seq2seq/train.py

```diff
@@ -0,0 +1,346 @@
+            sort_key="duration",
+            reverse=True,
+            key_max_value={"duration": hparams["avoid_if_longer_than"]},
+            key_min_value={"duration": hparams["avoid_if_smaller_than"]},
+        )
+        hparams["dataloader_options"]["shuffle"] = False
+    elif hparams["sorting"] == "random":
+        train_data = train_data.filtered_sorted(
+            key_max_value={"duration": hparams["avoid_if_longer_than"]},
+            key_min_value={"duration": hparams["avoid_if_smaller_than"]},
+        )
+
+    else:
+        raise NotImplementedError(
+            "sorting must be random, ascending or descending"
+        )
+
+    # We also sort the validation data so it is faster to validate.
+    valid_data = sb.dataio.dataset.DynamicItemDataset.from_csv(
+        csv_path=hparams["csv_valid"], replacements={"data_root": data_folder}
+    )
+    valid_data = valid_data.filtered_sorted(sort_key="duration", reverse=True)
+
+    # We also sort the test data so it is faster to validate.
+    test_data = sb.dataio.dataset.DynamicItemDataset.from_csv(
+        csv_path=hparams["csv_test"], replacements={"data_root": data_folder}
+    )
+    test_data = test_data.filtered_sorted(sort_key="duration", reverse=True)
+
+    datasets = [train_data, valid_data, test_data]
+
+    label_encoder = sb.dataio.encoder.CTCTextEncoder()
+
+    # 2. Define audio pipeline:
+    @sb.utils.data_pipeline.takes("wav", "start_seg", "end_seg")
+    @sb.utils.data_pipeline.provides("sig")
+    def audio_pipeline(wav, start_seg, end_seg):
+        start = int(float(start_seg) * hparams["sample_rate"])
+        stop = int(float(end_seg) * hparams["sample_rate"])
+        speech_segment = {"file": wav, "start": start, "stop": stop}
+        sig = sb.dataio.dataio.read_audio(speech_segment)
+        return sig
+
+    sb.dataio.dataset.add_dynamic_item(datasets, audio_pipeline)
+
+    # 3. Define text pipeline:
+    @sb.utils.data_pipeline.takes("char")
+    @sb.utils.data_pipeline.provides("char_list", "char_encoded")
+    def text_pipeline(char):
+        char_list = char.strip().split()
+        yield char_list
+        char_encoded = label_encoder.encode_sequence_torch(char_list)
+        yield char_encoded
+
+    sb.dataio.dataset.add_dynamic_item(datasets, text_pipeline)
+
+    # 4. Create a label encoder instead of a tokenizer for our tag list:
+    lab_enc_file = hparams["save_folder"] + "/labelencoder.txt"
+    label_encoder.load_or_create(
+        path=lab_enc_file,
+        from_didatasets=[train_data],
+        output_key="char_list",
+        special_labels={"blank_label": hparams["blank_index"]},
+        sequence_input=True,
+    )
+
+    # 5. Set output:
+    sb.dataio.dataset.set_output_keys(
+        datasets, ["id", "sig", "char_encoded"],
+    )
+
+    # 6. Make DataLoaders and shuffle if needed:
+    dataloader_train = torch.utils.data.DataLoader(
+        train_data,
+        batch_size=hparams["batch_size"],
+        num_workers=3,
```

Comment:
why fixing the num_workers here? Please consider using the hparam value instead.

Reply:
Correction done.